File: ApolloSwift/ApolloSwiftKit/ApolloBookService.swift
//
//  ApolloBookService.swift
//  ApolloSwift
//
//  Created by Christopher San Diego on 1/11/20.
//  Copyright © 2020 Christopher San Diego. All rights reserved.
//

import Apollo
import Foundation

public class ApolloBookService: BookService {
    
    private let client: ApolloClientProtocol
    private let queue: DispatchQueue
    
    public init(client: ApolloClientProtocol, queue: DispatchQueue = .main) {
        self.client = client
        self.queue = queue
    }
    
    public func list(completion: @escaping (Result<[Book], Error>) -> Void) {
        _ = client.fetch(query: BookListQuery(), cachePolicy: .returnCacheDataElseFetch, context: nil, queue: queue) { result in
            switch result {
            case .success(let graphQLResult):
                completion(.success(graphQLResult.data!.books.map { dto in
                    Book(id: dto.id, title: dto.title, author: dto.author)
                }))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    public func create(book: Book, completion: @escaping (Result<Bool?, Error>) -> Void) {
        _ = client.perform(mutation: BookCreateMutation(book: BookMutation(title: book.title, author: book.author)), context: nil, queue: queue) { result in
            switch result {
            case .success(let graphQLResult):
                completion(.success(graphQLResult.data?.createBook))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    public func get(id: Int, completion: @escaping (Result<Book, Error>) -> Void) {
        _ = client.fetch(query: BookGetQuery(id: id), cachePolicy: .returnCacheDataElseFetch, context: nil, queue: queue) { result in
            switch result {
            case .success(let graphQLResult):
                let dto = graphQLResult.data!.book
                completion(.success(Book(id: dto.id, title: dto.title, author: dto.author)))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    public func update(id: Int, book: Book, completion: @escaping (Result<Bool?, Error>) -> Void) {
        _ = client.perform(mutation: BookUpdateMutation(id: id, book: BookMutation(title: book.title, author: book.author)), context: nil, queue: queue) { result in
            switch result {
            case .success(let graphQLResult):
                completion(.success(graphQLResult.data?.updateBook))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    public func delete(id: Int, completion: @escaping (Result<Bool?, Error>) -> Void) {
        _ = client.perform(mutation: BookDeleteMutation(id: id), context: nil, queue: queue) { result in
            switch result {
            case .success(let graphQLResult):
                completion(.success(graphQLResult.data?.deleteBook))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
}

File: ApolloJvm/src/main/kotlin/com/github/csandiego/pocgrpahql/apollojvm/apollo/ApolloBookService.kt
package com.github.csandiego.pocgrpahql.apollojvm.apollo


import com.apollographql.apollo.ApolloMutationCall
import com.apollographql.apollo.ApolloQueryCall
import com.apollographql.apollo.coroutines.toDeferred
import com.github.csandiego.pocgrpahql.apollojvm.*
import com.github.csandiego.pocgrpahql.apollojvm.data.Book
import com.github.csandiego.pocgrpahql.apollojvm.service.BookService
import com.github.csandiego.pocgrpahql.apollojvm.type.BookMutation

class ApolloBookService(private val queryCallFactory: ApolloQueryCall.Factory, private val mutationCallFactory: ApolloMutationCall.Factory) : BookService {

    override suspend fun list(): List<Book> {
        return queryCallFactory.query(BookListQuery()).toDeferred().await().data()!!.books.map {
            Book(it.id, it.title, it.author)
        }
    }

    override suspend fun get(id: Int): Book {
        return queryCallFactory.query(BookGetQuery(id)).toDeferred().await().data()!!.book.let {
            Book(it.id, it.title, it.author)
        }
    }

    override suspend fun create(book: Book) {
        mutationCallFactory.mutate(BookCreateMutation(BookMutation(book.title, book.author))).toDeferred().await()
    }

    override suspend fun update(id: Int, book: Book) {
        mutationCallFactory.mutate(BookUpdateMutation(id, BookMutation(book.title, book.author))).toDeferred().await()
    }

    override suspend fun delete(id: Int) {
        mutationCallFactory.mutate(BookDeleteMutation(id)).toDeferred().await()
    }
}